imports/server/garbage-collection.ts

```diff
@@ -48,43 +48,103 @@ async function cleanup() {
     return;
   }
 
   Logger.info("Cleaning up dead servers", {
     deadServers: deadServers.join(","),
   });
-
-  // Run all hooks.
-  for (const f of globalGCHooks) {
-    await f(deadServers);
+  for (const deadServer of deadServers) {
+    await cleanupDeadServer(deadServer);
   }
-
-  // Delete the record of the server, now that we've cleaned up after it.
-  await Servers.removeAsync({ _id: { $in: deadServers } });
 }
 
 export async function cleanupDeadServer(id: string) {
+  // First mark that we're cleaning up the server so that if it wakes back up it
+  // knows to abort
+  const updated = await Servers.updateAsync(id, {
+    $set: { cleanupInProgressBy: serverId },
+  });
+  if (updated === 0) {
+    // Someone else is already cleaning this up
+    return;
+  }
+
   for (const f of globalGCHooks) {
-    await f([id]);
+    await f(id);
   }
   await Servers.removeAsync(id);
 }
 
+registerPeriodicCleanupHook(async (deadServer) => {
+  // If a server died while cleaning up another server, we want to unmark it as
+  // being cleaned up so that someone else picks it up
+  await Servers.updateAsync(
+    { cleanupInProgressBy: deadServer },
+    { $unset: { cleanupInProgressBy: "" } },
+  );
+});
+
 function periodic() {
   Meteor.setTimeout(periodic, 15000 + 15000 * Random.fraction());
   cleanup().catch((error) => {
     Logger.error("Error performing garbage-collection cleanup()", { error });
   });
 }
 
-// Defer the first run to give other startup hooks a chance to run
-Meteor.startup(() =>
-  Meteor.defer(() => {
-    Logger.info("New server starting", { serverId });
+Meteor.startup(() => {
+  // Defer the first run so that other startup hooks run first, but don't catch
+  // this async function's errors, because if it fails we want the failure to
+  // bubble up and abort the process
+  setImmediate(async () => {
+    Logger.info("New server starting", {
+      serverId,
+      pid: process.pid,
+      hostname: os.hostname(),
+    });
+
+    let handle: Meteor.LiveQueryHandle | undefined;
+    if (!Meteor.isAppTest) {
+      const aborted = () => {
+        Logger.error("Server record unexpectedly marked for deletion", {
+          serverId,
+          pid: process.pid,
+          hostname: os.hostname(),
+        });
+        process.exit(1);
+      };
+
+      handle = await Servers.find(serverId).observeChangesAsync({
+        changed(_, fields) {
+          if (fields.cleanupInProgressBy) {
+            aborted();
+          }
+        },
+        removed() {
+          aborted();
+        },
+      });
+    }
+
+    await ignoringDuplicateKeyErrors(async () => {
+      await Servers.insertAsync({
+        _id: serverId,
+        pid: process.pid,
+        hostname: os.hostname(),
+        updatedAt: new Date(),
+      });
+    });
+
     periodic();
-  }),
-);
+
+    onExit(async () => {
+      handle?.stop();
+      // Attempt to clean our own server record up synchronously on exit
+      await cleanupDeadServer(serverId);
```

Comment:
I think eager cleanup is still messy due to https://github.com/deathandmayhem/jolly-roger/blob/b95d4d36e3bd9a67bd598d65b9738f63d6afc3c4/imports/server/mediasoup-api.ts#L28-L55 trying to take over for the dead server.